### ISSUE TITLE
Refactor UniRL core architecture from an `Agent.observe` loop to a dataflow-oriented design

### ISSUE THREAD

**Issue by ilgyu-yi** [ilgyu-yi/unirl#10]
## Background

The current UniRL architecture is centered on the following loop:

```text
Env -> ObsAdapter -> Agent.act -> ActAdapter -> Env.step -> Agent.observe
```

This is reflected in the current codebase:

- `src/unirl/interfaces/agent.py`
- `src/unirl/interfaces/env.py`
- `src/unirl/interfaces/adapter.py`
- `src/unirl/interfaces/types.py`
- `src/unirl/system/system.py`
- `src/unirl/examples/*`
- `README.md`

In the current design:

- `System.run_episode()` owns the environment loop
- `Transition` is constructed inside `System`
- `Agent.observe(...)` is the place where transition ingestion and learning-related behavior are expected to happen

This design is simple, but it mixes several responsibilities into the `Agent`/`System` pair:

- rollout / environment interaction
- trajectory construction
- training data preparation
- parameter update triggering
- orchestration

This makes the current abstraction difficult to extend toward:

- PPO-style data pipelines without placing batch logic inside `Agent`
- replay-based learning
- AlphaZero / MuZero
- async / distributed actor-learner systems
- model-based RL
- RNN / belief-state agents that need resettable internal state but do not need `observe`

## Goal

Replace the current Agent-centric loop with a dataflow-oriented architecture built around explicit core contracts.

The new architecture should:

- move framework-level contracts from `interfaces/` into `core/`
- remove `Agent.observe(...)` from the core abstraction
- separate:
  - decision
  - rollout
  - batch sourcing
  - learning
  - orchestration
- keep `unirl` torch-free
- reserve `src/unirl/impl/` for concrete torch-based implementations
- make the intended design obvious from the package structure and interfaces alone

Backward compatibility is not a goal. This repository has not been released yet, so the architecture should be corrected directly rather than preserved through compatibility layers.

## Target architecture

The target design should model RL as:

```text
Obs -> Act -> Trajectory -> Batch -> Update
```

with the following control structure:

```text
Coordinator
  -> Rollout
       -> Agent
       -> Env
       -> Adapters
  -> BatchSource
  -> Learner
```

Important clarification:

- `Agent` is not a pipeline stage
- `Agent` is an operator used inside `Rollout`
- `Rollout` owns environment interaction and produces trajectories
- `Learner` updates the parameters used by the `Agent`
- `BatchSource` owns ingestion and sampling of training data

Parameter ownership should be understood as:

```text
Agent reads shared parameters
Learner updates shared parameters
```

The exact shared object does not need to be abstracted in core at this stage, but the README and docstrings must make this relationship explicit.

### Current structure

```text
src/unirl/
├── __init__.py
├── interfaces/
│   ├── __init__.py
│   ├── types.py
│   ├── env.py
│   ├── agent.py
│   └── adapter.py
├── system/
│   ├── __init__.py
│   └── system.py
├── registry/
│   ├── __init__.py
│   └── registry.py
├── config/
│   ├── __init__.py
│   └── loader.py
└── examples/
    ├── simple_env.py
    ├── simple_agent.py
    ├── simple_adapter.py
    └── configs/
        └── simple.yaml
```

### Target structure

```text
src/unirl/
├── __init__.py
├── core/
│   ├── __init__.py
│   ├── types.py
│   ├── env.py
│   ├── agent.py
│   ├── adapter.py
│   ├── rollout.py
│   ├── batch_source.py
│   ├── learner.py
│   └── coordinator.py
├── registry/
│   ├── __init__.py
│   └── registry.py
├── config/
│   ├── __init__.py
│   └── loader.py
├── impl/
│   ├── __init__.py
│   ├── agents/
│   ├── rollouts/
│   ├── batch_sources/
│   ├── learners/
│   └── models/
└── examples/
    ├── simple_env.py
    ├── simple_agent.py
    ├── simple_adapter.py
    └── configs/
        └── simple.yaml
```

### 1. Move `interfaces/` to `core/`

All framework-level protocols and types currently under `src/unirl/interfaces/` must move to `src/unirl/core/`.

This includes:

- `types.py`
- `env.py`
- `agent.py`
- `adapter.py`

All imports, re-exports, tests, and README references must be updated accordingly.

`src/unirl/interfaces/` should be removed, not retained as a shim.

### Current

```python
class Agent[AgentObsT, AgentActT](Protocol):
    def act(self, obs: AgentObsT) -> AgentActT: ...
    def observe(
        self,
        obs: AgentObsT,
        action: AgentActT,
        reward: float,
        next_obs: AgentObsT,
        terminated: bool,
        truncated: bool,
    ) -> None: ...
```

### Target

```python
class Agent[AgentObsT, AgentActT](Protocol):
    def act(self, obs: AgentObsT) -> AgentActT: ...
    def reset(self) -> None: ...
```

Requirements:

- remove `observe(...)` from the core interface
- add `reset()` for episode-boundary state reset
- `reset()` must support:
  - RNN hidden state reset
  - search tree reset
  - other agent-local episode state

The new `Agent` contract is:

- decision only
- stateful if needed
- no learning hook

### 3. Keep `Env` and adapters minimal

`Env`, `ObsAdapter`, and `ActAdapter` should be moved to `core/` with minimal semantic change.

The existing typed mapping remains valid:

```text
EnvObs -> AgentObs -> AgentAct -> EnvAct
```

### 4. Keep `StepResult`, narrow the role of `Transition`

`StepResult` remains a core type.

`Transition` may remain in `core/types.py` only as a minimal convenience dataclass if still useful for simple examples, but it must not be treated as the universal training record for all algorithms.

The issue should be implemented with the assumption that:

- algorithm-specific trajectory and batch types live in `impl/`
- core should not over-constrain those shapes

### 5. Add `Rollout`

Create `src/unirl/core/rollout.py`.

Suggested direction:

```python
class Rollout[
    EnvObsT,
    EnvActT,
    AgentObsT,
    AgentActT,
    TrajT,
](Protocol):
    def run_episode(
        self,
        env: Env[EnvObsT, EnvActT],
        agent: Agent[AgentObsT, AgentActT],
        obs_adapter: ObsAdapter[EnvObsT, AgentObsT],
        act_adapter: ActAdapter[AgentActT, EnvActT],
    ) -> TrajT:
        ...
```

Responsibilities:

- reset env
- call `agent.reset()`
- adapt observations/actions
- perform environment interaction
- construct and return a trajectory

This is intentionally close to the current `System.run_episode()` behavior so that the refactor is direct and traceable.

### 6. Add `BatchSource`

Create `src/unirl/core/batch_source.py`.

Suggested direction:

```python
class BatchSource[TrajT, BatchT](Protocol):
    def ingest(self, traj: TrajT) -> None:
        ...

    def sample(self, batch_size: int) -> BatchT:
        ...
```

Rationale:

- `BatchSource` is broader and more accurate than `Experience`
- it covers:
  - on-policy trajectory accumulation
  - replay buffers
  - target-building stores
  - in-memory or non-persistent implementations
- it does not over-emphasize persistence the way `DataStore` would

Important constraint:

- keep this contract minimal
- do not force a separate `process()` method in core
- implementations may preprocess on ingest, on sample, or both

### 7. Add `Learner`

Create `src/unirl/core/learner.py`.

Suggested direction:

```python
class Learner[BatchT](Protocol):
    def update(self, batch: BatchT) -> None:
        ...
```

Responsibilities:

- compute loss
- apply parameter update
- update the shared parameters used by `Agent`

This relationship must be described in code comments and README.

### 8. Add `Coordinator`

Create `src/unirl/core/coordinator.py`.

Suggested direction:

```python
class Coordinator(Protocol):
    def run(self) -> None:
        ...
```

Concrete implementation direction:

```python
class GenericCoordinator:
    def __init__(
        self,
        rollout,
        batch_source,
        learner,
        *,
        batch_size: int,
        rollout_per_iter: int,
        updates_per_iter: int,
    ):
        ...

    def run(self) -> None:
        while True:
            for _ in range(self.rollout_per_iter):
                traj = self.rollout.run_episode(...)
                self.batch_source.ingest(traj)

            for _ in range(self.updates_per_iter):
                batch = self.batch_source.sample(self.batch_size)
                self.learner.update(batch)
```

This is the reference control model for the refactor.

## System changes

`src/unirl/system/system.py` should no longer be the primary architecture anchor.

It should be rewritten or replaced so that the old `System + agent.observe` design is no longer central.

The implementation should not try to preserve the old behavior mechanically.

At minimum:

- remove any reliance on `Agent.observe`
- do not keep hidden learning logic inside `System`
- if `System` remains, it should become a thin convenience wrapper around rollout-style logic, not the architectural center

If keeping `system/` introduces unnecessary confusion, it is acceptable to remove or substantially repurpose it.

## `impl/` policy

Create the `src/unirl/impl/` package skeleton in this issue.

Required subpackages:

```text
src/unirl/impl/
├── __init__.py
├── agents/
├── rollout/
├── batch_source/
├── learners/
└── models/
```

Rules:

- `core/` defines contracts
- `impl/` provides concrete implementations
- `impl/` may require torch
- no duplicate protocol layer under `impl/`
- do not create `impl/interfaces.py`

This issue is not required to fully populate `impl/`, but it must establish the layout so that future torch-based PRs have an obvious home.

## Examples changes

The current examples under `src/unirl/examples/` must remain minimal and reference-only.

They should be updated to match the new core contracts.

Specifically:

- `simple_agent.py` must implement:
  - `act(...)`
  - `reset()`
- it must not implement `observe(...)`
- examples should remain torch-free
- examples should continue to illustrate:
  - typed env
  - typed adapters
  - typed agent
  - YAML registration flow where applicable

## Registry and config changes

Current registry/config code is built around constructing the existing `System`.

This issue must update those paths so they remain consistent with the new architecture.

At minimum:

- update imports from `interfaces` to `core`
- ensure registry naming and decorators still work
- ensure YAML/config-driven assembly still has a valid path forward

Do not leave stale assumptions about `Agent.observe` in registry/config code.

## README changes

Update the root README to match the new architecture.

Required changes:

1. replace all `interfaces/` references with `core/`
2. remove `Agent.observe(...)` from the core interface section
3. remove the old Agent-centric episode-loop diagram
4. replace it with a dataflow-oriented description
5. document the new roles:
   - Agent
   - Rollout
   - BatchSource
   - Learner
   - Coordinator
6. explain clearly that:
   - Agent reads shared parameters
   - Learner updates them
7. ensure examples and package layout sections match the new code layout

The updated README must describe the actual architecture after this refactor, not the pre-refactor design.

## Concrete implementation checklist

- [ ] create `src/unirl/core/`
- [ ] move current protocol/type modules from `interfaces/` into `core/`
- [ ] update imports and top-level re-exports
- [ ] remove `src/unirl/interfaces/`
- [ ] remove `Agent.observe(...)` from the core contract
- [ ] add `Agent.reset()`
- [ ] add `core/rollout.py`
- [ ] add `core/batch_source.py`
- [ ] add `core/learner.py`
- [ ] add `core/coordinator.py`
- [ ] refactor or replace `system/system.py` so it no longer encodes the old `observe`-driven design
- [ ] update examples to the new interfaces
- [ ] create `src/unirl/impl/` package skeleton
- [ ] update registry/config imports and assumptions
- [ ] update README to reflect the new architecture
- [ ] ensure `pyright --strict` passes
- [ ] ensure tests pass after the refactor

## Acceptance criteria

This issue is complete when all of the following are true:

1. `core/` exists and is the only home for framework-level contracts
2. `interfaces/` has been removed
3. `Agent.observe(...)` no longer exists in the core API
4. `Agent.reset()` exists in the core API
5. `Rollout`, `BatchSource`, `Learner`, and `Coordinator` exist as first-class core abstractions
6. the codebase no longer treats `Agent` as the place where learning happens
7. `impl/` exists as the home for future torch-based implementations
8. the README accurately describes the new architecture
9. the issue can be implemented without needing to infer missing architectural intent

## Summary

Replace the current:

```text
System + Agent.observe
```

architecture with:

```text
Coordinator
  -> Rollout
       -> Agent
       -> Env
       -> Adapters
  -> BatchSource
  -> Learner
```

and move all framework contracts into `src/unirl/core/`.

+ Chore: Follow generic protocol style PEP 695
